File: packages/bouncer-py/examples/transactional_claim.py
from __future__ import annotations

import bouncer


def main() -> None:
    db = bouncer.open("app.sqlite3")
    db.bootstrap()

    with db.transaction() as tx:
        tx.execute(
            "CREATE TABLE IF NOT EXISTS jobs (payload TEXT NOT NULL)"
        )
        tx.execute(
            "INSERT INTO jobs(payload) VALUES (?)",
            ["run scheduler tick"],
        )

        claim = tx.claim("scheduler", "worker-a", ttl_ms=30_000)
        if claim.acquired:
            print(f"acquired token {claim.lease.token}")
        else:
            print(f"busy: currently owned by {claim.current.owner}")
            tx.rollback()


if __name__ == "__main__":
    main()

File: packages/bouncer/examples/transactional_claim.rs
use std::time::Duration;

use litelease::{Bouncer, ClaimResult};
use rusqlite::params;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut db = Bouncer::open("app.sqlite3")?;
    db.bootstrap()?;

    let tx = db.transaction()?;
    tx.conn().execute(
        "CREATE TABLE IF NOT EXISTS jobs (payload TEXT NOT NULL)",
        [],
    )?;
    tx.conn().execute(
        "INSERT INTO jobs(payload) VALUES (?1)",
        params!["run scheduler tick"],
    )?;

    match tx.claim("scheduler", "worker-a", Duration::from_secs(30))? {
        ClaimResult::Acquired(lease) => {
            println!("acquired token {}; committing", lease.token);
            tx.commit()?;
        }
        ClaimResult::Busy(current) => {
            println!("busy: owned by {}; rolling back", current.owner);
            tx.rollback()?;
        }
    }

    Ok(())
}
